src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -0,0 +1,770 @@
+				case WP_HTML_Processor_State::INSERTION_MODE_IN_BODY:
+					return $this->step_in_body();
+
+				default:
+					throw new WP_HTML_Unsupported_Exception( 'Cannot parse outside of the IN BODY insertion mode.' );
+			}
+		} catch ( WP_HTML_Unsupported_Exception $e ) {
+			if ( defined( 'DEBUG' ) && 1 !== preg_match( '~Cannot process [A-Z0-9-]+ element.~', $e->getMessage() ) ) {
+				throw $e;
+			}
+
+			/*
+			 * Exceptions are used in this class to escape deep call stacks that
+			 * otherwise might involve messier calling and return conventions.
+			 */
+			return false;
+		}
+	}
+
+	/**
+	 * Computes the HTML breadcrumbs for the currently-matched node, if matched.
+	 *
+	 * @return WP_HTML_Token[]|null
+	 */
+	public function get_breadcrumbs() {
+		if ( ! $this->get_tag() ) {
+			return null;
+		}
+
+		$breadcrumbs = array();
+		foreach ( $this->state->stack_of_open_elements->walk_down() as $stack_item ) {
+			$breadcrumbs[] = $stack_item;
+		}
+
+		return $breadcrumbs;
+	}
+
+	/**
+	 * Parses next element in the 'in body' insertion mode.
+	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inbody
+	 *
+	 * @return boolean Whether an element was found.
+	 * @throws WP_HTML_Unsupported_Exception
+	 */
+	private function step_in_body() {
+		$tag_name = $this->get_tag();
+		$op_sigil = $this->is_tag_closer() ? '-' : '+';
+		$op       = "{$op_sigil}{$tag_name}";
+
+		switch ( $op ) {
+			/*
+			 * > A start tag whose tag name is one of: "address", "article", "aside",
+			 * > "blockquote", "center", "details", "dialog", "dir", "div", "dl",
+			 * > "fieldset", "figcaption", "figure", "footer", "header", "hgroup",
+			 * > "main", "menu", "nav", "ol", "p", "search", "section", "summary", "ul"
+			 */
+			case '+BLOCKQUOTE':
+			case '+DIV':
+			case '+FIGCAPTION':
+			case '+FIGURE':
+			case '+P':
+				if ( $this->state->stack_of_open_elements->has_p_in_button_scope() ) {
+					$this->close_a_p_element();
+				}
+
+				$this->insert_html_element( $this->current_token );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is one of: "address", "article", "aside", "blockquote",
+			 * > "button", "center", "details", "dialog", "dir", "div", "dl", "fieldset",
+			 * > "figcaption", "figure", "footer", "header", "hgroup", "listing", "main",
+			 * > "menu", "nav", "ol", "pre", "search", "section", "summary", "ul"
+			 */
+			case '-BLOCKQUOTE':
+			case '-DIV':
+			case '-FIGCAPTION':
+			case '-FIGURE':
+				if ( ! $this->state->stack_of_open_elements->has_element_in_scope( $tag_name ) ) {
+					// Ignore the token.
+					return $this->step();
+				}
+
+				$this->generate_implied_end_tags();
+				$this->state->stack_of_open_elements->pop_until( $tag_name );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "p"
+			 */
+			case '-P':
+				if ( ! $this->state->stack_of_open_elements->has_p_in_button_scope() ) {
+					$this->insert_html_element( $this->current_token );
+				}
+
+				$this->close_a_p_element();
```

Comment:
It took me a minute to figure out it doesn't close both the `a` element and a `p` element, how about this instead?

```suggestion
				$this->close_p_element();
```

Reply:
torn on this one because of the correspondence to the spec terminology. if we call it `close_p_element()` then I think it carries a different connotation, that we assume we're currently directly inside a `P` element, which is not what's happening. if we call it `close_nearest_p_element()` then I think that matches better the meaning, but diverges from the spec.

what if we call it `close_a_P_element()`?